packages/element/src/create-interpolate-element.js

```diff
@@ -0,0 +1,235 @@
+		return potentialElement;
+	}
+	const [ mapItem ] = conversionMap.slice( 0, 1 );
+	const [ searchString, conversionConfig ] = mapItem;
+
+	/**
+	 * This short circuits the process if the conversion map has an invalid config.
+	 */
+	if ( ! searchString || ! conversionConfig ) {
+		return potentialElement;
+	}
+
+	const match = getMatchFromString(
+		potentialElement,
+		searchString,
+		conversionConfig
+	);
+
+	// if there is no match for this string, then that means it is not an element
+	// so just return as is to be used as a direct child.
+	if ( match === null ) {
+		return potentialElement;
+	}
+
+	// if the full match returned equals the potential element, then we know
+	// we can create the element and restart the conversion on any children if
+	// necessary.
+	if ( match[ 0 ] === potentialElement ) {
+		// remove this item from the conversion map because it's no longer needed.
+		conversionMap.shift();
+
+		if ( getHasPropValue( conversionConfig ) ) {
+			// if value is a react element, then need to wrap in Fragment with a key
+			// to prevent key warnings.
+			if ( isValidElement( conversionConfig.value ) ) {
+				keyIndex++;
+				return <Fragment key={ keyIndex }>{ conversionConfig.value }</Fragment>;
+			}
+			return conversionConfig.value;
+		}
+		keyIndex++;
+		return getHasChildren( conversionConfig ) ?
+			createElement(
+				conversionConfig.tag,
+				{ ...conversionConfig.props, key: keyIndex },
+				recursiveCreateElement( match[ 1 ], conversionMap )
+			) :
+			createElement(
+				conversionConfig.tag,
+				{ ...conversionConfig.props, key: keyIndex }
+			);
+	}
+
+	// still here, so we need to split on the full match and loop through each
+	return flatMap(
+		potentialElement.split( getSplitRegEx( match[ 0 ] ) )
+			.filter( ( value ) => !! value ),
+		( element ) => {
+			return recursiveCreateElement( element, conversionMap );
+		}
+	);
+};
+
+/**
+ * This function creates an interpolated element from a passed in string with
+ * specific tags matching how the string should be converted to an element via
+ * the conversion map value.
+ *
+ * @example
+ * For example, for the given string:
+ *
+ * "This is a <span%1>string</span%1> with <a%1>a link</a%1>, a self-closing
+ * %1$s tag and a plain value %2$s"
+ *
+ * You would have something like this as the conversionMap value:
+ *
+ * ```js
+ * {
+ *     'span%1': { tag: CustomComponent, props: {} },
+ *     'a%1': { tag: 'a', props: { href: 'https://github.com' } },
+ *     '%1$s': { tag: CustomComponentB, props: {} },
+ *     '%2$s': { value: 'custom value' },
+ * }
+ * ```
+ *
+ * @param {string}  interpolatedString  The interpolation string to be parsed.
+ * @param {Object} conversionMap        The map used to convert the string to
+ *                                      a react element.
+ *
+ * @return {Element}  A react element.
+ */
+const createInterpolateElement = ( interpolatedString, conversionMap ) => {
+	keyIndex = -1;
```

Comment:
I'd love it if we could avoid keeping this kind of state in the module. I realise, given the kind of data being treated and the synchronous nature of this operation, that the risk of tainting this state is low, but fundamentally `keyIndex` is a value we should be able to let recurse without external state.

Reply:
The entirety of `create-interpolate-element.js` is the module and `keyIndex` is external to the specific function but not the module.  It was written this way because (at least with the current recursion logic) passing through `keyIndex` as an argument won't work due to the nesting handling. The iteration of `keyIndex` at the parent level needs to increment if children update it.

What suggestion would you have for an alternative approach?

Reply:
> It was written this way because (at least with the current recursion logic) passing through `keyIndex` as an argument won't work due to the nesting handling.

Yeah, I suspected this would be why.

> What suggestion would you have for an alternative approach?

I don't have a concrete approach, although intuitively maybe this would work if `keyIndex` weren't sequential, but a vector of sorts to represent branching/nesting — e.g. if, instead of just going from `0` to `Inf`, it would support values like `'1'`, `'1.0.2'`… so that nested calls of `recursiveCreateElement` can act without reporting data back to their callers.

But this only an intuition, I haven't tried any of it. And I wouldn't block the PR over it. But does it sound like a viable approach?

Reply:
> Although intuitively maybe this would work if keyIndex weren't sequential, but a vector of sorts to represent branching/nesting

This would be nice  as it would connect branches with parents more visibly via the key value, but there would still be the issue of how would the value be bumped for parents (i.e. 1.0 -> 2.0) when there's a new branch?  I think I'd run into the same issue that I had with straight incrementation.  

One possibility would be to pass an object along (via an argument) instead of a primitive.  As a reference it would be mutable and thus parents would always have the latest iterative value.  But I feel that from a risk vector, that's not much different than the current implementation.